kinder/pkg/cri/nodes/containerd/alterhelper.go

```diff
@@ -120,3 +121,24 @@ func Commit(containerID, targetImage string) error {
 
 	return cmd.Run()
 }
+
+// GetSandboxImage returns the sandbox image used by the container runtime
+func GetSandboxImage(bc *bits.BuildContext) (string, error) {
+	cmd := "crictl -D=false info -o go-template --template {{.config.sandboxImage}}"
```

Comment:
seems to work:
```
time="10:28:25" level=debug msg="Running: docker exec kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f bash -c crictl -D=false info -o go-template --template {{.config.sandboxImage}}"
time="10:28:25" level=info msg="Detected sandbox image registry.k8s.io/pause:3.7"
time="10:28:25" level=debug msg="Running: docker pull registry.k8s.io/pause:3.9"
time="10:28:26" level=debug msg="Running: docker save -o=/tmp/kinder-image-path1320318010/pause.tar registry.k8s.io/pause:3.9"
time="10:28:26" level=debug msg="Running: docker cp /tmp/kinder-image-path1320318010/pause.tar kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f:/kind/images"
time="10:28:26" level=debug msg="Running: docker exec kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f ctr --namespace=k8s.io image import /kind/images/pause.tar --no-unpack"
time="10:28:26" level=debug msg="Running: docker exec kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f rm /kind/images/pause.tar"
time="10:28:26" level=debug msg="Running: docker pull registry.k8s.io/etcd:3.5.9-0"
time="10:28:33" level=debug msg="Running: docker save -o=/tmp/kinder-image-path1320318010/etcd.tar registry.k8s.io/etcd:3.5.9-0"
time="10:28:39" level=debug msg="Running: docker cp /tmp/kinder-image-path1320318010/etcd.tar kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f:/kind/images"
time="10:28:40" level=debug msg="Running: docker exec kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f ctr --namespace=k8s.io image import /kind/images/etcd.tar --no-unpack"
time="10:28:45" level=debug msg="Running: docker exec kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f rm /kind/images/etcd.tar"
time="10:28:45" level=debug msg="Running: docker pull registry.k8s.io/coredns/coredns:v1.11.1"
time="10:28:48" level=debug msg="Running: docker save -o=/tmp/kinder-image-path1320318010/coredns.tar registry.k8s.io/coredns/coredns:v1.11.1"
time="10:28:50" level=debug msg="Running: docker cp /tmp/kinder-image-path1320318010/coredns.tar kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f:/kind/images"
time="10:28:50" level=debug msg="Running: docker exec kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f ctr --namespace=k8s.io image import /kind/images/coredns.tar --no-unpack"
time="10:28:51" level=debug msg="Running: docker exec kind-build-f2e926d3-1350-4755-97bc-c5ba4239c76f rm /kind/images/coredns.tar"
time="10:28:51" level=debug msg="Running: docker pull registry.k8s.io/pause:3.7"
```
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubeadm/2933/pull-kubeadm-kinder-upgrade-latest/1701904818869637120

but instead of detecting it, we could also just setup containerd to match what kubeadm version X needs.
that would avoid the double pull of e.g. above example 3.7 and 3.9.

https://kubernetes.io/docs/setup/production-environment/container-runtimes/#override-pause-image-containerd
i think we already had one area of kinder where we update the containerd config.

here it is:
https://github.com/kubernetes/kubeadm/blob/c1258e7214bae23cd6fff88aff6470e649c2e22e/kinder/pkg/cri/nodes/containerd/alterhelper.go#L49-L54

might as well we can use a TOML library:
https://github.com/pelletier/go-toml
or just string read/write file with string replace.

but please comment if you think that downloading two images is better than updating config.toml.
i don't mind it either way for pause...for e.g. etcd it's a different story.

Reply:
> but instead of detecting it, we could also just setup containerd to match what kubeadm version X needs.

+1 for this approach @neolit123 
And the code has been updated to set up the containerd to match what kubeadm version X needs.

> but has to happen for init/upgrade, might be a bit tricky

I think it's enough to just prepare for the init phase because our e2e tests for `kubeadm init` have covered all the supported Kubernetes versions, and the compatibility has been well-verified.

Also, we should verify that `kubeadm upgrade` works well without updating the sandbox image of runtime, which seems to be a very common scenario.

Reply:
during upgrade kinder downloads a list of images after consulting the 'config images list' command. the pause there could differ from the already set during init pause in the containerd config but that is ok i think.